src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -35,54 +36,28 @@ use syntax::ast::LitKind;
 use syntax::feature_gate::{UnstableFeatures, feature_err, emit_feature_err, GateIssue};
 use syntax_pos::{Span, DUMMY_SP};
 
-use std::fmt;
-use rustc_data_structures::sync::Lrc;
 use std::usize;
+use std::fmt;
 
 use transform::{MirPass, MirSource};
 use super::promote_consts::{self, Candidate, TempState};
 
-bitflags! {
-    // Borrows of temporaries can be promoted only if
-    // they have none of these qualifications, with
-    // the exception of `STATIC_REF` (in statics only).
-    struct Qualif: u8 {
-        // Constant containing interior mutability (UnsafeCell).
-        const MUTABLE_INTERIOR  = 1 << 0;
-
-        // Constant containing an ADT that implements Drop.
-        const NEEDS_DROP        = 1 << 1;
-
-        // Function argument.
-        const FN_ARGUMENT       = 1 << 2;
-
-        // Not constant at all - non-`const fn` calls, asm!,
-        // pointer comparisons, ptr-to-int casts, etc.
-        const NOT_CONST         = 1 << 3;
-
-        // Refers to temporaries which cannot be promoted as
-        // promote_consts decided they weren't simple enough.
-        const NOT_PROMOTABLE    = 1 << 4;
+pub const MUT_INTERIOR_BITS: u8 = 1 << 0;
+pub const NEEDS_DROP_BITS: u8 = 1 << 1;
+pub const NOT_CONST_BITS: u8 = 1 << 2;
 
-        // Const items can only have MUTABLE_INTERIOR
-        // and NOT_PROMOTABLE without producing an error.
-        const CONST_ERROR       = !Qualif::MUTABLE_INTERIOR.bits &
-                                  !Qualif::NOT_PROMOTABLE.bits;
+fn qualif_bits((mut_interior, needs_drop, not_const): (bool, bool, bool)) -> u8 {
+    let mut bits = 0;
+    if mut_interior {
+        bits |= MUT_INTERIOR_BITS;
     }
-}
-
-impl<'a, 'tcx> Qualif {
-    /// Remove flags which are impossible for the given type.
-    fn restrict(&mut self, ty: Ty<'tcx>,
-                tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                param_env: ty::ParamEnv<'tcx>) {
-        if ty.is_freeze(tcx, param_env, DUMMY_SP) {
-            *self = *self - Qualif::MUTABLE_INTERIOR;
-        }
-        if !ty.needs_drop(tcx, param_env) {
-            *self = *self - Qualif::NEEDS_DROP;
-        }
+    if needs_drop {
+        bits |= NEEDS_DROP_BITS;
     }
+    if not_const {
+        bits |= NOT_CONST_BITS;
+    }
```

Comment:
You can simplify this by using an enum `Bits` and then:
```rust
(mut_interior as u8) << (Bits::MutInterior as u32) |
(needs_drop as u8) << (Bits::NeedsDrop as u32) |
(not_const as u8) << (Bits::NotConst as u32)
```

Reply:
Good point!